src/data/bucket/fill_bucket.js

```diff
@@ -65,18 +68,49 @@ class FillBucket implements Bucket {
     }
 
     populate(features: Array<IndexedFeature>, options: PopulateParameters) {
+        const patterns = options.patternDependencies;
+        this.features = [];
+
+        for (const layer of this.layers) {
+            const fillPattern = layer.paint.get('fill-pattern');
+            const images = fillPattern.property.getPossibleOutputs();
+            for (const image of images) {
+                patterns[image] = true;
+            }
+        }
+
         for (const {feature, index, sourceLayerIndex} of features) {
             if (this.layers[0]._featureFilter(new EvaluationParameters(this.zoom), feature)) {
+
                 const geometry = loadGeometry(feature);
-                this.addFeature(feature, geometry, index);
+                const fillFeature: BucketFeature = {
+                    sourceLayerIndex: sourceLayerIndex,
+                    index: index,
+                    geometry: geometry,
+                    properties: feature.properties,
+                    type: feature.type
+                };
+
+                if (typeof feature.id !== 'undefined') {
+                    fillFeature.id = feature.id;
+                }
+
+                this.features.push(fillFeature);
```

Comment:
this follows the pattern that is also used in symbol layout. we could split it into separate steps, but this way avoids repeating the work  of parsing features out of the vector layers here:
https://github.com/mapbox/mapbox-gl-js/blob/2d0c4bc9872b51efefa77d41cf71ed806c8eb3c8/src/source/worker_tile.js#L77-L94

Reply:
I was going to say, since we're restricting `*-pattern` properties to values where the possible outputs are contained as literals within the expression, we can take advantage of that to calculate the needed pattern images without reference to the actual features. But it looks like you've lifted that restriction in the latest revision. Do you remember the original rationale for it? On the one hand, it's nice that it would allow this optimization, but on the other, I can definitely envision someone wanting to use e.g. `"fill-pattern": ["get", "pattern-name"]`, since that's been so useful for `icon-image`.

Reply:
I think the original reason for the restriction was to avoid the per-feature evaluation before the layout step, but yes it has been lifted and I agree, the ["get", "pattern-name"] use-case is probably going to be one users will appreciate and use.